### PR TITLE
Rework raw ident suggestions

### LINKED ISSUE #66126: Raw ident suggestion should only trigger if the code is otherwise well-formed
Right now the raw ident suggestion when using a reserved keyword somewhere where it is not allowed is context free, leading to misleading output:

https://twitter.com/isislovecruft/status/1191768296590315520

The parser should perform some lookahead to see if it could actually continue parsing if a raw identifier had been supplied in its place and if so, only then suggest raw idents.

### PR DESCRIPTION
**Pull request by estebank**
Use heuristics to determine whethersuggesting raw identifiers is
appropriate.

Account for raw identifiers when printing a path in a `use` suggestion.

Fix #66126.

### CODE REVIEW

File: src/librustc_parse/parser/diagnostics.rs [hunk 1]
```diff
@@ -225,8 +225,21 @@ impl<'a> Parser<'a> {
             self.token.span,
             &format!("expected identifier, found {}", self.this_token_descr()),
         );
+        let valid_follow = &[
+            TokenKind::Eq,
+            TokenKind::Colon,
+            TokenKind::Comma,
+            TokenKind::Semi,
+            TokenKind::ModSep,
+            TokenKind::OpenDelim(token::DelimToken::Brace),
```
Comment (cramertj) [inline src/librustc_parse/parser/diagnostics.rs]: I think this should also include `Bracket`-- e.g.:

```rust
macro_rules! async { () => {} }
```

Comment (cramertj) [reply src/librustc_parse/parser/diagnostics.rs]: (but it is a heuristic, and as such it may be better not to suggest changing `async { ...  }`)

**Comment by cramertj** [comment]
r=me with travis fixed

File: src/libsyntax_pos/symbol.rs [hunk 2]
```diff
@@ -1011,7 +1011,15 @@ pub mod sym {
 }
 
 impl Symbol {
-    fn is_used_keyword_2018(self) -> bool {
+    pub fn is_used_keyword(self) -> bool {
+        self.is_used_keyword_2015() || self.is_used_keyword_2018()
+    }
+
+    pub fn is_used_keyword_2015(self) -> bool {
+        self >= kw::As && self <= kw::While
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: Could you remove the `symbol.rs` changes?
`is_used_keyword` needs a span to work correctly.

If there's no available span, you can make a dummy spanned ident with `Ident::with_dummy_span` a the call point.

File: src/librustc_parse/parser/diagnostics.rs [hunk 3]
```diff
@@ -225,8 +225,21 @@ impl<'a> Parser<'a> {
...
+            TokenKind::OpenDelim(token::DelimToken::Paren),
+            TokenKind::CloseDelim(token::DelimToken::Brace),
+            TokenKind::CloseDelim(token::DelimToken::Paren),
+        ];
```
Comment (petrochenkov) [inline src/librustc_parse/parser/diagnostics.rs]: Code like this is pure technical debt, and my preferred solution would be to remove the diagnostic (https://github.com/rust-lang/rust/issues/66126#issuecomment-549956637), especially given that raw identifiers are a compatibility feature that should never be *recommended* in general.

File: src/librustc_resolve/lib.rs [hunk 4]
```diff
@@ -2893,15 +2893,16 @@ fn names_to_string(names: &[Name]) -> String {
         if i > 0 {
             result.push_str("::");
         }
+        if name.is_used_keyword() && name.can_be_raw() {
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]:
```suggestion
        if Ident::with_dummy_span(name).is_raw_guess() {
```

File: src/libsyntax_pos/symbol.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax_pos/symbol.rs]: done

File: src/librustc_resolve/lib.rs (hunk 4 above)
Comment (estebank) [reply src/librustc_resolve/lib.rs]: done

File: src/librustc_parse/parser/diagnostics.rs (hunk 3 above)
Comment (estebank) [reply src/librustc_parse/parser/diagnostics.rs]: I think that with this new restriction we'll only be presenting this suggestion in the very rare cases where the _intent_ was there.

**Pull request merged**